### PR TITLE
Select push app in preferences

### LINKED ISSUE #674: Add a proper way to set external application for pushing
I searched for 10 minutes how to change the external application for pushing from emacs to vim, when I finally found the drop-down next to the tool bar button.

Please add a proper way of changing this setting through a menu action. I think it's not good style to have an action only available through the tool bar.

### PR DESCRIPTION
**Pull request by calixtus**
As asked by @tobiasdiez, here is my alternative proposal, this closes #4991 by merging.
fixes #674 closes #3942 and closes #4368

PrefPushToApp

This branch is build up on the other branch, including the other commits, this time by intention.
I hope you like it. Have a good night!

----

- [x] Change in CHANGELOG.md described
- [ ] Tests created for changes
- [x] Manually tested changed features in running JabRef
- [x] Screenshots added in PR description (for bigger UI changes)
- [x] Ensured that [the git commit message is a good one](https://github.com/joelparkerhenderson/git_commit_message)
- [x] Check documentation status (Issue created for outdated help page at [help.jabref.org](https://github.com/JabRef/help.jabref.org/issues)?)

### CODE REVIEW

**Comment by calixtus** [comment]
Ready to review

**Comment by calixtus** [comment]
I added some eye-candy, you will be charmed. ;-)
select push application

**Comment by Siedlerchr** [comment]
Looks good to me!

File: src/main/java/org/jabref/gui/preferences/ExternalTab.java [hunk 1]
```diff
@@ -197,24 +215,19 @@ public Node getBuilder() {
...
     @Override
     public void setValues() {
 
         emailSubject.setText(prefs.get(JabRefPreferences.EMAIL_SUBJECT));
         openFoldersOfAttachedFiles.setSelected(prefs.getBoolean(JabRefPreferences.OPEN_FOLDERS_OF_ATTACHED_FILES));
 
+        String pushToApplicationName = prefs.get(JabRefPreferences.PUSH_TO_APPLICATION);
+        List<PushToApplication> pushToApplications = frame.getPushToApplicationsManager().getApplications();
+        pushToApplicationComboBox.setValue(
+                pushToApplications.stream()
```
Comment (tobiasdiez) [inline src/main/java/org/jabref/gui/preferences/ExternalTab.java]: Maybe extract this logic to a method `getActivePushToApplication(PushToApplicationsManager manager)` in `JabRefPreferences`?

File: src/main/java/org/jabref/gui/preferences/ExternalTab.java [hunk 2]
```diff
@@ -244,6 +257,10 @@ public void setValues() {
     public void storeSettings() {
         prefs.put(JabRefPreferences.EMAIL_SUBJECT, emailSubject.getText());
         prefs.putBoolean(JabRefPreferences.OPEN_FOLDERS_OF_ATTACHED_FILES, openFoldersOfAttachedFiles.isSelected());
+        if (pushToApplicationComboBox.getValue().getApplicationName() != JabRefPreferences.PUSH_TO_APPLICATION) {
```
Comment (tobiasdiez) [inline src/main/java/org/jabref/gui/preferences/ExternalTab.java]: Similarily, I would extract part of this code to `setActivePushToApplication(PushToApplication)` method in `JabRefPreferences`.

File: src/main/java/org/jabref/gui/preferences/ExternalTab.java [hunk 3]
```diff
@@ -276,6 +293,21 @@ private void updateExecuteConsoleButtonAndFieldEnabledState() {
         consoleCommand.setDisable(!executeConsole.isSelected());
     }
 
+    private void showPushToApplicationSettings() {
+        PushToApplicationsManager manager = frame.getPushToApplicationsManager();
+        PushToApplication selectedApplication = manager.getApplicationByName(pushToApplicationComboBox.getValue().getApplicationName());
```
Comment (tobiasdiez) [inline src/main/java/org/jabref/gui/preferences/ExternalTab.java]: `selectedApplication = pushToApplicationComboBox.getValue()` should suffice

File: src/main/java/org/jabref/gui/preferences/ExternalTab.java [hunk 4]
```diff
@@ -276,6 +293,21 @@ private void updateExecuteConsoleButtonAndFieldEnabledState() {
...
+        PushToApplicationSettings settings = manager.getSettings(selectedApplication);
+
+        DialogPane dialogPane = new DialogPane();
+        dialogPane.setContent(settings.getJFXSettingPane(pushToApplicationComboBox.getValue().getApplicationName()));
+
+        dialogService.showCustomDialogAndWait(Localization.lang("App settings"), dialogPane, ButtonType.OK, ButtonType.CANCEL).ifPresent(btn -> {
```
Comment (tobiasdiez) [inline src/main/java/org/jabref/gui/preferences/ExternalTab.java]: Here too, please no abbreviation.

File: src/main/java/org/jabref/gui/push/PushToApplicationAction.java [hunk 5]
```diff
@@ -25,25 +26,29 @@
...
 
-    private final PushToApplication operation;
     private final StateManager stateManager;
     private final DialogService dialogService;
 
+    private PushToApplication application;
+
     public PushToApplicationAction(StateManager stateManager, PushToApplicationsManager pushToApplicationsManager, DialogService dialogService) {
-        this.operation = pushToApplicationsManager.getLastUsedApplication(Globals.prefs);
+        this.application = pushToApplicationsManager.getApplicationByName(Globals.prefs.get(JabRefPreferences.PUSH_TO_APPLICATION));
```
Comment (tobiasdiez) [inline src/main/java/org/jabref/gui/push/PushToApplicationAction.java]: reuse method I proposed to introduce above

File: src/main/java/org/jabref/gui/push/PushToApplicationSettings.java [hunk 6]
```diff
@@ -23,24 +23,24 @@ public PushToApplicationSettings(DialogService dialogService) {
         this.dialogService = dialogService;
     }
 
-    public GridPane getJFXSettingPane(int n) {
-        switch (n) {
-            case 0:
+    public GridPane getJFXSettingPane(String applicationName) {
```
Comment (tobiasdiez) [inline src/main/java/org/jabref/gui/push/PushToApplicationSettings.java]: What about tests of the form `application is instanceof(...)`, then you don't need to work with "magic" strings.

File: src/main/java/org/jabref/gui/push/PushToApplicationsManager.java [hunk 7]
```diff
@@ -40,11 +62,30 @@ public PushToApplicationSettings getSettings(PushToApplication application) {
...
-                           .filter(application -> application.getApplicationName().equals(appSelected))
+                           .filter(application -> application.getApplicationName().equals(applicationName))
                            .findAny()
                            .orElse(applications.get(0));
     }
+
+    public void updateApplicationAction() {
+        final ActionFactory factory = new ActionFactory(Globals.getKeyPrefs());
+
+        action.updateApplication(getApplicationByName(Globals.prefs.get(JabRefPreferences.PUSH_TO_APPLICATION)));
```
Comment (tobiasdiez) [inline src/main/java/org/jabref/gui/push/PushToApplicationsManager.java]: same again: reuse method

File: src/main/java/org/jabref/gui/push/PushToApplicationSettings.java (hunk 6 above)
Comment (calixtus) [reply src/main/java/org/jabref/gui/push/PushToApplicationSettings.java]: I found it very hard to understand the code in PushToApplicationSettings, so i took the liberty, to refactor this class a little bit.

**Comment by calixtus** [comment]
Ok, should be ready again.

**Comment by Siedlerchr** [comment]
Looks good, please check if you can select the push to application multiple times after each other without closing the prefs dialog
Select lyx and then again select Texstudio for example. 
If it works all is good

**Comment by calixtus** [comment]
Works fine.

Christoph <notifications@github.com> schrieb am Di., 4. Juni 2019, 20:01:

**Comment by tobiasdiez** [comment]
Looks very good to me now. Thanks again!

**Comment by calixtus** [comment]
You are welcome. This was fun - again! Let's push JabRaf forward towards 5.0. ;-)

**Pull request merged**